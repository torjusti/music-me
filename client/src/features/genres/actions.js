export const setGenreSelected = (genre, selected) => ({
  type: 'SET_GENRE_SELECTED',
  payload: {
    genre,
    selected,
  },
});
