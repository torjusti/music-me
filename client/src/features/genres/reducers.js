const initialState = {
  availableGenres: [],
  selectedGenres: [],
};

const genres = (state = initialState, action) => {
  switch (action.type) {
    case 'SET_GENRES':
      return { ...state, availableGenres: action.payload.genres };

    case 'SET_SELECTED_GENRE':
      return {
        ...state,
        selectedGenres: action.payload.selected ?
          [...state.selectedGenres, action.payload.genre] :
          state.selectedGenres.filter(genre => genre !== action.payload.genre),
      };

    default:
      return state;
  }
};

export default genres;
