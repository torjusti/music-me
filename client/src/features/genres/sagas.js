import { call, put, takeLatest } from 'redux-saga/effects';
import { requestData } from '../../common/api';

/**
 * Saga which retrieves available genres for filtering from the server..
 */
function* fetchGenres() {
  // Send a request to the server for genres.
  const response = yield call(requestData, `/genres`);

  if (response.error) {
    // Show an error notification if errros occur.
    yield put({ type: 'FETCH_ERROR' });
  }

  if (response.data) {
    // Send an action to update the available genres if
    // the genres are retrieved correctly from the server.
    yield put({ type: 'SET_GENRES', payload: { genres: response.data } });
  }
}

function* genresSaga() {
  // Run the above saga when FETCH_GENRES actions are retrieved.
  // If another FETCH_GENRES action is retrieved before the last
  // one is finished, the previous is cancelled and the last one starts.
  yield takeLatest(['FETCH_GENRES'], fetchGenres);
}

export default genresSaga;
