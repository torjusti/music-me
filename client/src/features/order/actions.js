/**
 * Sets the column on which we are sorting.
 */
export const setColumn = orderBy => ({
  type: 'SET_COLUMN',
  payload: { orderBy },
});

/**
 * Toggles the sorting direction on the currently selected column.
 */
export const toggleDirection = () => ({
  type: 'TOGGLE_DIRECTION',
});

/**
 * Signifies that we should stop sorting on the currently selected column.
 */
export const clearOrder = () => ({
  type: 'CLEAR_ORDER',
});
