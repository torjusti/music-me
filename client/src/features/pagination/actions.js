/**
 * Sets the pagination page which is being shown in the table.
 */
export const setPage = page => ({
  type: 'SET_PAGE',
  payload: { page },
});
