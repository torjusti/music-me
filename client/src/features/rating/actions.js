export const setRatingSelected = selectedRating => ({
  type: 'SET_SELECTED_RATING',
  payload: {
    selectedRating,
  },
});

export const toggleRatingEnabled = () => ({
  type: 'TOGGLE_RATING_ENABLED',
});
