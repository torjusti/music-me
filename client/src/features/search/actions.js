export const setQuery = query => ({
  type: 'SET_QUERY',
  payload: { query },
});

export const clearQuery = () => ({
  type: 'CLEAR_QUERY',
});
