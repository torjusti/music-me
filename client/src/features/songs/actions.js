/**
 * Action which eventually sends a rating for a song to the server.
 */
export const sendSongRating = (id, rating) => ({
  type: 'SEND_SONG_RATING',
  payload: { id, rating },
});
