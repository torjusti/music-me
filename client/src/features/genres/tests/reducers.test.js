import genres from '../reducers';
import * as actions from '../actions';

describe('genres reducer', () => {
  it('should return an empty array initially', () => {
    expect(genres(undefined, {})).toEqual({availableGenres: [], selectedGenres: []});
  });

  it('should set a genre', () => {
    expect(
      genres(
        undefined,
        actions.setGenreSelected(
          ['setGenreSelected'],
          true,
        ),
      ),
    ).toMatchSnapshot();
  });

  it('should set no genre', () => {
    expect(
      genres(
        undefined,
        actions.setGenreSelected(
          ['setGenreSelected'],
          false
        ),
      ),
    ).toMatchSnapshot();
  });

  it('should set none genre selected', () => {
    expect(
      genres(
        undefined,
        actions.setGenreSelected(
          [],
          true,
        ),
      ),
    ).toMatchSnapshot();
  });

  it('should select all genre', () => {
    expect(
      genres(
        undefined,
        actions.setGenreSelected(
          ['setGenreSelected', 'genre'],
          true,
        ),
      ),
    ).toMatchSnapshot();
  });
});
