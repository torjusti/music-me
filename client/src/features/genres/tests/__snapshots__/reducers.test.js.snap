// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`genres reducer should select all genre 1`] = `
Object {
  "availableGenres": Array [],
  "selectedGenres": Array [
    Array [
      "setGenreSelected",
      "genre",
    ],
  ],
}
`;

exports[`genres reducer should set a genre 1`] = `
Object {
  "availableGenres": Array [],
  "selectedGenres": Array [
    Array [
      "setGenreSelected",
    ],
  ],
}
`;

exports[`genres reducer should set no genre 1`] = `
Object {
  "availableGenres": Array [],
  "selectedGenres": Array [],
}
`;

exports[`genres reducer should set none genre selected 1`] = `
Object {
  "availableGenres": Array [],
  "selectedGenres": Array [
    Array [],
  ],
}
`;
